### PR TITLE
Recursive implementation

### PR DESCRIPTION
**Pull request by janusheide**

### CODE REVIEW

File: src/endian/big_endian.hpp [hunk 1]
```diff
@@ -8,277 +8,185 @@
...
-{
-    static void put(ValueType value, uint8_t* buffer)
-    {
-        static_assert(sizeof(ValueType) <= 1, "ValueType too big.");
-        assert(buffer != nullptr);
-        *buffer = value;
-    }
+// Helper to convet floating point type into identically sized unsigned integer
+template<class Type>
+struct FloatType
```
Comment (mortenvp) [inline src/endian/big_endian.hpp]: lower-case is our style for classes and structs

File: src/endian/big_endian.hpp [hunk 2]
```diff
@@ -8,277 +8,185 @@
...
-        static_assert(std::is_unsigned<ValueType>::value, "Must be unsigned");
-        assert(buffer != nullptr);
-        assert(value <= 0x0000FFFFFFFFFFFF);
-        buffer[5] = ((value >> 0) & 0xFF);
-        buffer[4] = ((value >> 8) & 0xFF);
-        buffer[3] = ((value >> 16) & 0xFF);
-        buffer[2] = ((value >> 24) & 0xFF);
-        buffer[1] = ((value >> 32) & 0xFF);
-        buffer[0] = ((value >> 40) & 0xFF);
+        static_assert(Bytes > sizeof(ValueType) / 2,
```
Comment (mortenvp) [inline src/endian/big_endian.hpp]: What is the purpose of the division?

File: src/endian/big_endian.hpp [hunk 3]
```diff
@@ -8,277 +8,185 @@
...
 };
 
-template<class ValueType>
-struct big<ValueType, 3>
+
+template<class ValueType, uint8_t Bytes>
+struct big_impl
 {
-    static void put(ValueType value, uint8_t* buffer)
+    static void put(ValueType& value, uint8_t* buffer)
```
Comment (mortenvp) [inline src/endian/big_endian.hpp]: assert(buffer != nullptr);

Comment (janusheide) [reply src/endian/big_endian.hpp]: this check has already been made in the public API - in all the non-detail functions

File: src/endian/big_endian.hpp [hunk 4]
```diff
@@ -1,284 +1,142 @@
...
-                (((ValueType) buffer[3]) << 8)  |
-                (((ValueType) buffer[4]) << 0);
+        value |= ((ValueType) *buffer);
     }
 };
 
-template<class ValueType>
-struct big<ValueType, 6>
+
+// Helper to delegate to the appropiate specialization depednign on the type
```
Comment (jpihl) [inline src/endian/big_endian.hpp]: Spalling

File: src/endian/little_endian.hpp [hunk 5]
```diff
@@ -1,286 +1,140 @@
...
+        value |= ((ValueType) *buffer);
     }
 };
 
-template<class ValueType>
-struct little<ValueType, 5>
+// Helper to delegate to the appropiate specialization depednign on the type
+// @TODO remove these wrappers when we have CXX17 support and "if constexpr"
+template<class ValueType, uint8_t Bytes,
+         bool isUnsigened = std::is_unsigned<ValueType>::value,
```
Comment (jpihl) [inline src/endian/little_endian.hpp]: I would start these isSomething with a capital letter.

**Pull request merged**